libs/SmartStore/src/com/salesforce/androidsdk/smartstore/store/DBOpenHelper.java

```diff
@@ -492,34 +519,55 @@ public boolean removeExternalBlobsDirectory(String soupTableName) {
 	}
 
 	/**
-	 * Re-encrypts the files on external storage with the new key. If external storage is not enabled for any table in the db, this operation is ignored.
+	 * Changes the encryption key on the database.
+	 *
+	 * @param db Database object.
+	 * @param oldKey Old encryption key.
+	 * @param newKey New encryption key.
+	 */
+	public static synchronized void changeKey(SQLiteDatabase db, String oldKey, String newKey) {
+		db.query("PRAGMA rekey = '" + newKey + "'");
+	}
+
+	/**
+	 * Re-encrypts the files on external storage with the new key. If external storage is not
+	 * enabled for any table in the db, this operation is ignored.
 	 *
 	 * @param db DB containing external storage (if applicable).
 	 * @param oldKey Old key with which to decrypt the existing data.
 	 * @param newKey New key with which to encrypt the existing data.
 	 */
-	public static void reEncryptAllFiles(SQLiteDatabase db, String oldKey, String newKey) {
-		StringBuilder path = new StringBuilder(db.getPath()).append(EXTERNAL_BLOBS_SUFFIX);
-		File dir = new File(path.toString());
+	public static synchronized void reEncryptAllFiles(SQLiteDatabase db, String oldKey, String newKey) {
+		final File dir = new File(db.getPath() + EXTERNAL_BLOBS_SUFFIX);
 		if (dir.exists()) {
-			File[] tables = dir.listFiles();
+			final File[] tables = dir.listFiles();
 			if (tables != null) {
-				for (File table : tables) {
-					File[] blobs = table.listFiles();
+				for (final File table : tables) {
+					final File[] blobs = table.listFiles();
 					if (blobs != null) {
-						for (File blob : blobs) {
-							StringBuilder json = new StringBuilder();
-							String result = null;
+						for (final File blob : blobs) {
+							final StringBuilder json = new StringBuilder();
+							String result;
 							try {
-								BufferedReader br = new BufferedReader(new FileReader(blob));
+								final BufferedReader br = new BufferedReader(new FileReader(blob));
 								String line;
 								while ((line = br.readLine()) != null) {
 									json.append(line).append('\n');
 								}
 								br.close();
-								result = Encryptor.decrypt(json.toString(), oldKey);
+
+								/*
+								 * If the key length is 24, then it's the old key (16 bytes for the
+								 * key and 8 bytes for the IV. If the key length is 44, then it's the
+								 * new key (32 bytes for the key and 12 bytes for the IV).
+								 */
+								if (oldKey.getBytes().length == 24) {
```

Comment:
@wmathurin Added a check here based on key size - `legacyDecrypt()` is applied to the old key size and `decrypt()` is applied to the new key size. Brought back the test as well.